GVFS/GVFS.UnitTests/Common/Database/PlaceholdersTableTests.cs

```diff
@@ -0,0 +1,435 @@
+        private const byte PathTypeExpandedFolder = 2;
+        private const byte PathTypePossibleTombstoneFolder = 3;
+        private const string DefaultSha = "SHA1";
+
+        [TestCase]
+        public void ConstructorTest()
+        {
+            Mock<IGVFSConnectionPool> mockConnectionPool = new Mock<IGVFSConnectionPool>(MockBehavior.Strict);
+            PlaceholdersTable placeholders = new PlaceholdersTable(mockConnectionPool.Object);
+            mockConnectionPool.VerifyAll();
+        }
+
+        [TestCase]
+        public void CreateTableTest()
+        {
+            Mock<IDbCommand> mockCommand = new Mock<IDbCommand>(MockBehavior.Strict);
+            mockCommand.SetupSet(x => x.CommandText = "CREATE TABLE IF NOT EXISTS [Placeholders] (path TEXT PRIMARY KEY, pathType TINYINT NOT NULL, sha char(40) ) WITHOUT ROWID;");
+            mockCommand.Setup(x => x.ExecuteNonQuery()).Returns(1);
+            PlaceholdersTable.CreateTable(mockCommand.Object);
+            mockCommand.VerifyAll();
+        }
+
+        [TestCase]
+        public void CountTest()
+        {
+            this.TestPlaceholders(
+                (placeholders, mockCommand) =>
+                {
+                    mockCommand.SetupSet(x => x.CommandText = "SELECT count(path) FROM Placeholders;");
+                    mockCommand.Setup(x => x.ExecuteScalar()).Returns(0);
+                    placeholders.Count().ShouldEqual(0);
+                });
+        }
+
+        [TestCase]
+        public void GetAllFilePathsWithNoResults()
+        {
+            this.TestPlaceholdersWithReader(
+               (placeholders, mockCommand, mockReader) =>
+               {
+                   mockReader.Setup(x => x.Read()).Returns(false);
+                   mockCommand.SetupSet(x => x.CommandText = "SELECT path FROM Placeholders WHERE pathType = 0;");
+
+                   HashSet<string> filePaths = placeholders.GetAllFilePaths();
+                   filePaths.ShouldNotBeNull();
+                   filePaths.Count.ShouldEqual(0);
+               });
+        }
+
+        [TestCase]
+        public void GetAllFilePathsTest()
+        {
+            this.TestPlaceholdersWithReader(
+               (placeholders, mockCommand, mockReader) =>
+               {
+                   int readCalls = 0;
+                   mockReader.Setup(x => x.Read()).Returns(() =>
+                   {
+                       ++readCalls;
+                       return readCalls == 1;
+                   });
+
+                   mockReader.Setup(x => x.GetString(0)).Returns(DefaultPath);
+                   mockCommand.SetupSet(x => x.CommandText = "SELECT path FROM Placeholders WHERE pathType = 0;");
+
+                   HashSet<string> filePaths = placeholders.GetAllFilePaths();
+                   filePaths.ShouldNotBeNull();
+                   filePaths.Count.ShouldEqual(1);
+                   filePaths.ShouldContain(x => x == DefaultPath);

```

Comment:
Should these `filePaths.ShouldContain(x => x == <thing>)` instead be `filePaths.Contains(<thing>).ShouldBeTrue()`? The `ShouldContain()` enumerates the set (it only knows its an `IEnumerable<T>`).